File: php/Day5.php
<?php
    require_once('Utils.php');

    function Day5_Part1() {
        $x1 = array();
        $y1 = array();
        $x2 = array();
        $y2 = array();

        foreach (read_file_lines("../data/day5.txt") as $row) {
            array_push($x1, (int) explode(",", explode(" -> ", $row)[0])[0]);
            array_push($y1, (int) explode(",", explode(" -> ", $row)[0])[1]);
            array_push($x2, (int) explode(",", explode(" -> ", $row)[1])[0]);
            array_push($y2, (int) explode(",", explode(" -> ", $row)[1])[1]);
        }

        $max_x = max(max($x1), max($x2));
        $max_y = max(max($y1), max($y2));

        $vents = array();
        for ($y = 0; $y < $max_y + 1; $y++) {
            array_push($vents, array());
            for ($x = 0; $x < $max_x + 1; $x++)
            array_push($vents[$y], 0);
        }

        for ($index = 0; $index < count($x1); $index++) {
            $x_one = $x1[$index];
            $y_one = $y1[$index];
            $x_two = $x2[$index];
            $y_two = $y2[$index];

            if ($x_one > $x_two) {
                $tmp = $x_one;
                $x_one = $x_two;
                $x_two = $tmp;
            }
            if ($y_one > $y_two) {
                $tmp = $y_one;
                $y_one = $y_two;
                $y_two = $tmp;
            }

            if ($x_one == $x_two)
                for ($y = $y_one; $y < $y_two + 1; $y++)
                    $vents[$y][$x_one]++;

            else if ($y_one == $y_two)
                for ($x = $x_one; $x < $x_two + 1; $x++)
                    $vents[$y_one][$x]++;
        }
    
        $counter = 0;
    
        foreach ($vents as $row) {
            foreach ($row as $cell) {
                if ($cell >= 2) $counter++;
            }
        }

        echo "Day:  5 | Part: 1 | Result: " . $counter . "<br>";
    }

    function Day5_Part2() {
        $x1 = array();
        $y1 = array();
        $x2 = array();
        $y2 = array();

        foreach (read_file_lines("../data/day5.txt") as $row) {
            array_push($x1, (int) explode(",", explode(" -> ", $row)[0])[0]);
            array_push($y1, (int) explode(",", explode(" -> ", $row)[0])[1]);
            array_push($x2, (int) explode(",", explode(" -> ", $row)[1])[0]);
            array_push($y2, (int) explode(",", explode(" -> ", $row)[1])[1]);
        }

        $max_x = max(max($x1), max($x2));
        $max_y = max(max($y1), max($y2));

        $vents = array();
        for ($y = 0; $y < $max_y + 1; $y++) {
            array_push($vents, array());
            for ($x = 0; $x < $max_x + 1; $x++)
            array_push($vents[$y], 0);
        }

        for ($index = 0; $index < count($x1); $index++) {
            $x_one = $x1[$index];
            $y_one = $y1[$index];
            $x_two = $x2[$index];
            $y_two = $y2[$index];

            if ($x_one == $x_two) {                    
                if ($y_one > $y_two) {
                    $tmp = $y_one;
                    $y_one = $y_two;
                    $y_two = $tmp;
                }

                for ($y = $y_one; $y < $y_two + 1; $y++)
                    $vents[$y][$x_one]++;
            } else if ($y_one == $y_two) {
                if ($x_one > $x_two) {
                    $tmp = $x_one;
                    $x_one = $x_two;
                    $x_two = $tmp;
                }
                
                for ($x = $x_one; $x < $x_two + 1; $x++)
                    $vents[$y_one][$x]++;
            } else {
                if ($x_one > $x_two) {
                    $tmp = $x_one;
                    $x_one = $x_two;
                    $x_two = $tmp;
                    $tmp = $y_one;
                    $y_one = $y_two;
                    $y_two = $tmp;
                }

                
                for ($o = 0; $o < $x_two - $x_one + 1; $o++)
                    $vents[$y_one < $y_two ? $y_one + $o : $y_one - $o][$x_one + $o]++;
            }
        }
    
        $counter = 0;
    
        foreach ($vents as $row) {
            foreach ($row as $cell) {
                if ($cell >= 2) $counter++;
            }
        }

        echo "Day:  5 | Part: 2 | Result: " . $counter . "<br>";
    }

    Day5_Part1();
    Day5_Part2();

File: go/Day5.go
package main

import (
	"aoc2021/Utils"
	"fmt"
	"strings"
)

func Day5_Part1() {
	lines := Utils.ReadFile("../data/day5.txt")

	x1 := make([]int, 0)
	y1 := make([]int, 0)
	x2 := make([]int, 0)
	y2 := make([]int, 0)

	var max_x int
	var max_y int

	for line_id := range lines {

		points := strings.Split(lines[line_id], " -> ")
		one := strings.Split(points[0], ",")
		two := strings.Split(points[1], ",")

		var x_one = Utils.StrToInt(one[0], 10)
		var y_one = Utils.StrToInt(one[1], 10)
		var x_two = Utils.StrToInt(two[0], 10)
		var y_two = Utils.StrToInt(two[1], 10)

		x1 = append(x1, x_one)
		y1 = append(y1, y_one)
		x2 = append(x2, x_two)
		y2 = append(y2, y_two)

		if x_one > max_x {
			max_x = x_one
		}
		if x_two > max_x {
			max_x = x_two
		}
		if y_one > max_y {
			max_y = y_one
		}
		if y_two > max_y {
			max_y = y_two
		}
	}

	vents := make([][]int, max_y+1)
	for y := 0; y <= max_y; y++ {
		vents[y] = make([]int, max_x+1)
		for x := 0; x <= max_x; x++ {
			vents[y][x] = 0
		}
	}

	for index := 0; index < len(x1); index++ {
		var x_one = x1[index]
		var y_one = y1[index]
		var x_two = x2[index]
		var y_two = y2[index]

		if x_one > x_two {
			x_one, x_two = x_two, x_one
		}
		if y_one > y_two {
			y_one, y_two = y_two, y_one
		}

		if x_one == x_two {
			for y := y_one; y <= y_two; y++ {
				vents[y][x_one]++
			}
		} else if y_one == y_two {
			for x := x_one; x <= x_two; x++ {
				vents[y_one][x]++
			}
		}
	}

	var counter = 0

	for i := 0; i < len(vents); i++ {
		for cell := 0; cell < len(vents[i]); cell++ {
			if vents[i][cell] >= 2 {
				counter++
			}
		}
	}

	fmt.Printf("Day:  5 | Part: 1 | Result: %d\n", counter)
}

func Day5_Part2() {
	lines := Utils.ReadFile("../data/day5.txt")

	x1 := make([]int, 0)
	y1 := make([]int, 0)
	x2 := make([]int, 0)
	y2 := make([]int, 0)

	var max_x int
	var max_y int

	for line_id := range lines {

		points := strings.Split(lines[line_id], " -> ")
		one := strings.Split(points[0], ",")
		two := strings.Split(points[1], ",")

		var x_one = Utils.StrToInt(one[0], 10)
		var y_one = Utils.StrToInt(one[1], 10)
		var x_two = Utils.StrToInt(two[0], 10)
		var y_two = Utils.StrToInt(two[1], 10)

		x1 = append(x1, x_one)
		y1 = append(y1, y_one)
		x2 = append(x2, x_two)
		y2 = append(y2, y_two)

		if x_one > max_x {
			max_x = x_one
		}
		if x_two > max_x {
			max_x = x_two
		}
		if y_one > max_y {
			max_y = y_one
		}
		if y_two > max_y {
			max_y = y_two
		}
	}

	vents := make([][]int, max_y+1)
	for y := 0; y <= max_y; y++ {
		vents[y] = make([]int, max_x+1)
		for x := 0; x <= max_x; x++ {
			vents[y][x] = 0
		}
	}

	for index := 0; index < len(x1); index++ {
		var x_one = x1[index]
		var y_one = y1[index]
		var x_two = x2[index]
		var y_two = y2[index]

		if x_one == x_two {
			if y_one > y_two {
				y_one, y_two = y_two, y_one
			}

			for y := y_one; y <= y_two; y++ {
				vents[y][x_one]++
			}
		} else if y_one == y_two {
			if x_one > x_two {
				x_one, x_two = x_two, x_one
			}

			for x := x_one; x <= x_two; x++ {
				vents[y_one][x]++
			}
		} else {
			if x_one > x_two {
				x_one, x_two = x_two, x_one
				y_one, y_two = y_two, y_one
			}

			var mult = 1
			if y_one > y_two {
				mult = -1
			}

			for o := 0; o <= x_two-x_one; o++ {
				vents[y_one+(o*mult)][x_one+o]++
			}
		}
	}

	var counter = 0

	for i := 0; i < len(vents); i++ {
		for cell := 0; cell < len(vents[i]); cell++ {
			if vents[i][cell] >= 2 {
				counter++
			}
		}
	}

	fmt.Printf("Day:  5 | Part: 1 | Result: %d\n", counter)
}
